Restructure NEST 3.0 guides (#1929)
nest/nest-simulator: doc/userdoc/guides/nest2_to_nest3/nest3_features/parametrization.rst

  |                                                    | Create spatially positioned nodes distributed freely  |
  |     nest.spatial.free(pos, extent=None,            | in space with dimensions given by `pos` or            |
  |         edge_wrap=False, num_dimensions=None)      | `num_dimensions`.                                     |
  |                                                    |                                                       |
  +----------------------------------------------------+-------------------------------------------------------+

  .. code-block:: ipython

    grid_nodes = nest.Create('iaf_psc_alpha', positions=nest.spatial.grid(shape=[10, 8]))
    nest.PlotLayer(grid_nodes);

.. image:: ../../../static/img/NEST3_23_0.png
  :width: 500px

.. code-block:: ipython

    free_nodes = nest.Create('iaf_psc_alpha', 100,
                             positions=nest.spatial.free(nest.random.uniform(min=0., max=10.),
                                                         num_dimensions=2))
    nest.PlotLayer(free_nodes);

.. image:: ../../../static/img/NEST3_24_0.png
  :width: 500px

> After you have created your spatially distributed nodes, you can use `spatial` property to set

Reviewer: ```suggestion
After you have created your spatially distributed nodes, you can use the `spatial` property to set
```

  |                                  |                                                                         |
  |     nest.spatial.distance        | | Distance between two nodes. Can only be used when connecting.         |
  +----------------------------------+-------------------------------------------------------------------------+
  |  ::                              |                                                                         |
  |                                  |                                                                         |
  |     nest.spatial.distance.x      |                                                                         |
  |     nest.spatial.distance.y      | | Distance on the x, y and z axis between the source and target neuron. |
  |     nest.spatial.distance.z      | | Can only be used when connecting.                                     |
  +----------------------------------+-------------------------------------------------------------------------+

  These parameters represent positions of neurons or distances between two
  neurons. To set node parameters, only the node position can be used. The

>   others can only be used when connecting.

Reviewer: ```suggestion
  others can be used when connecting.
```

    # Also available:
    #   - nest.math.sin()
    #   - nest.math.cos()

    spatial_nodes.set({'V_m': parameter})

    node_pos = np.array(nest.GetPosition(spatial_nodes))
    node_pos[:,1]
    v_m = spatial_nodes.get('V_m');

    fig, ax = pyplot.subplots(figsize=(12, 6))
    ax.plot(node_pos[:,0], v_m, '.', ms=6.5)
    ax.set_xlabel('Node position on x-axis')
    ax.set_ylabel('V_m');

.. image:: ../../../static/img/NEST3_27_0.png

.. _logic:

Clipping, redraw, and conditionals
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  +----------------------------------------------------+-----------------------------------------------------+
  | Parameter                                          | Description                                         |
  +====================================================+=====================================================+
  | ::                                                 |                                                     |
  |                                                    |                                                     |
  |     nest.math.min(x, value)                        | If a value from the Parameter is above a threshold, |

>   |                                                    | the value is replaced with the value of the         |

Reviewer: ```suggestion
  |                                                    | x, the value is replaced with the value of the      |
```

  |                                                    | threshold.                                          |
  +----------------------------------------------------+-----------------------------------------------------+
  | ::                                                 |                                                     |
  |                                                    |                                                     |

>   |     nest.math.max(x, value)                        | If a value from the parameter is beneath a          |
>   |                                                    | threshold, the value is replaced with the value of  |
>   |                                                    | the threshold.                                      |

Reviewer: ```suggestion
  |     nest.math.max(x, value)                        | If a value from the parameter is below a            |
  |                                                    | threshold x, the value is replaced with the value      |
  |                                                    | of the threshold.                                   |
```

  |                                                    | if a suitable value is not found after a certain    |
  |                                                    | number of redraws.                                  |
  +----------------------------------------------------+-----------------------------------------------------+
  | ::                                                 |                                                     |
  |                                                    |                                                     |
  |     nest.logic.conditional(x, val_true, val_false) | Given a condition, yields one value or another      |
  |                                                    | based on if the condition evaluates to true or      |
  |                                                    | false.                                              |
  +----------------------------------------------------+-----------------------------------------------------+

Note that ``x`` is a ``nest.Parameter``.

The ``nest.math.min()`` and ``nest.math.max()`` functions are used to clip
a parameter. Essentially they work like the standard ``min()`` and

> ``max()`` functions, ``nest.math.min()`` yielding the smallest of two
> values, and ``nest.math.max()`` yielding the largest of two values.

Reviewer: ```suggestion
``max()`` functions, ``nest.math.min()`` yielding the smaller of two
values, and ``nest.math.max()`` yielding the larger of two values.
```

::

    # This yields values between 0.0 and 0.5, where values from the

>     # distribution that are above 0.5 gets set to 0.5.

Reviewer: ```suggestion
    # distribution that are above 0.5 get set to 0.5.
```

    nest.math.min(nest.random.uniform(), 0.5)

    # This yields values between 0.5 and 1.0, where values from the

>     # distribution that are below 0.5 gets set to 0.5.

Reviewer: ```suggestion
    # distribution that are below 0.5 get set to 0.5.
```

    nest.math.max(nest.random.uniform(), 0.5)

    # This yields values between 0.2 and 0.7, where values from the

>     # distribution that are smaller than 0.2 or larger than 0.7 gets

Reviewer: ```suggestion
    # distribution that are smaller than 0.2 or larger than 0.7 get
```

    # redrawn from the distribution.
    nest.math.redraw(nest.random.uniform(), min=0.2, max=0.7)

The ``nest.logic.conditional()`` function works like an ``if``/``else``
statement. Three arguments are required:

- The first argument is a condition.

> - The second argument is the resulting value or parameter evalued if the

Reviewer: ```suggestion
- The second argument is the resulting value or parameter evaluated if the
```

  condition evaluates to true.

> - The third argument is the resulting value or parameter evalued if the

Reviewer: ```suggestion
- The third argument is the resulting value or parameter evaluated if the
```

  |     nest.random.lognormal(mean=0.0, std=1.0)     | Draws samples based on a                   |
  |                                                  | lognormal distribution.                    |
  +--------------------------------------------------+--------------------------------------------+

For every value to be generated, samples are drawn from a distribution. The distribution uses
NEST's random number generator, and are therefore thread-safe. Note that
arguments can be passed to each of the distributions above to control the parameters of the
distribution.

.. code-block:: ipython

    n = nest.Create('iaf_psc_alpha', 10000, {'V_m': nest.random.normal(mean=-60.0, std=10.0)})

    node_ids = n.global_id
    v_m = n.get('V_m')
    fig, ax = pyplot.subplots(figsize=(12, 6),
                           gridspec_kw={'width_ratios':
                                        [3, 1]},
                           ncols=2,
                           sharey=True)
    ax[0].plot(node_ids, v_m, '.', alpha=0.5, ms=3.5)
    ax[0].set_xlabel('Node_ID');
    ax[1].hist(v_m, bins=40, orientation='horizontal');
    ax[1].set_xlabel('num. nodes');
    ax[0].set_ylabel('V_m');

> .. image:: ../../../_static/img/NEST3_13_0.png

Reviewer: Is this resolved?

  |                                                            |                                                                 |
  | ::                                                         | ::                                                              |
  |                                                            |                                                                 |
  |     kernel = {"gaussian": {"p_center": 1.0, "sigma": 1.0}} |     param = nest.spatial_distributions.gaussian(                |
  |                                                            |         nest.spatial.distance, p_center=1.0, std_deviation=1.0) |
  |                                                            |                                                                 |
  +------------------------------------------------------------+-----------------------------------------------------------------+

.. code-block:: ipython

    N = 21
    middle_node = N//2

    positions = nest.spatial.free([[x, 0.] for x in np.linspace(0, 1.0, N)])
    spatial_nodes = nest.Create('iaf_psc_alpha', positions=positions)

>     parameter = nest.spatial_distributions.exponential(nest.spatial.distance, beta_Ca=0.15)

Reviewer 2: ```suggestion
    parameter = nest.spatial_distributions.exponential(nest.spatial.distance, beta=0.15)
```
